### DISCUSSION TITLE
Huge grafana queries

### DISCUSSION THREAD

**Discussion by SBlechmann** [crate/crate#13963] [Q&A]
Hey there,

I hope this is not obsolete.. I am not quite the crate pro. Hope you can help me out here :) 

Since I have faced similar issues like https://github.com/crate/crate/issues/9748, I was just was just wondering if there is a best practice on how to work with huge queries generated by grafana.

So I have lots of devices sending data to crate every second. Also, I have a dashboard that visualizes lot's of data from different tables. If I now want to visualize data, let's say for a month, I run into the same issue as #9748 . I also saw the discussion about using the LIMIT parameter... so how to perform such a query now when I actually want to visualize that much data?

**Comment by hlcianfagna** [comment]
Hi @SBlechmann ,
#9748 is a bit old now, it concerned an old version of CrateDB, and if I am reading the thread correctly it had something to do with the cluster being too busy to create replicas for the new data that was arriving. That is not something I would expect you to observe on a recent version, with an appropriately sized cluster, if the tables ingesting the data have a reasonable [sharding and partitioning strategy](https://community.crate.io/t/sharding-and-partitioning-guide-for-time-series-data/737), and the queries being run are sensible.
There may also be some other reasons why replicas are not getting created, so it is always worth checking `sys.allocations`
A dashboard typically fits on a couple of screens, so from a user perspective it does not display a lot of data but a summary of a lot of data, you may for instance have a gauge that displays the average of millions of values, but the gauge itself is only displaying one value.
I imagine in your case you may have something like a line plot for a long period of time, there may be billions of records for this period of time, but on screen the line plot may just be something like 500 pixels wide, so we really only need 500 data points not millions of them, but we need the right 500 data points, not just the latest, so a `LIMIT` is usually not what we are looking for.
I would suggest some form of downsampling.
Take a look at 
https://community.crate.io/t/resampling-time-series-data-with-date-bin/1009
and
https://community.crate.io/t/advanced-downsampling-with-the-lttb-algorithm/1287

**Comment by SBlechmann** [comment]
Hey @hlcianfagna,
I am still using some older version (4.6.6) so I might need to update soon :s 
Yet, an error like this occured and the shard locked after trying to query lots of data.
The provided approaches seem promising, I'll take a look, thanks 👍
